File: rxrecycler-adapter/src/main/java/com/minimize/android/rxrecycleradapter/RxDataSource.java
package com.minimize.android.rxrecycleradapter;

import android.databinding.ViewDataBinding;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;

import java.util.Collections;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;

/**
 * Created by ahmedrizwan on 26/12/2015.
 */
public class RxDataSource<DataType>
{

    private List<DataType> mDataSet;
    private RxAdapter mRxAdapter;
    private RxAdapterForTypes<DataType> mRxAdapterForTypes;

    public RxDataSource(List<DataType> dataSet)
    {
        this.mDataSet = dataSet;
    }

    /***
     * Call this when binding with a single Item-Type
     *
     * @param recyclerView RecyclerView instance
     * @param item_layout Layout id
     * @param <LayoutBinding> ViewDataBinding Type for the layout
     * @return Observable for binding viewHolder
     */
    public <LayoutBinding extends ViewDataBinding> Observable<SimpleViewHolder<DataType, LayoutBinding>> bindRecyclerView(
            @NonNull final RecyclerView recyclerView, @NonNull @LayoutRes final int item_layout)
    {
        mRxAdapterForTypes = null;
        mRxAdapter = new RxAdapter(item_layout, mDataSet);
        recyclerView.setAdapter(mRxAdapter);
        return mRxAdapter.asObservable();
    }

    /***
     * Call this if you need access to the Adapter!
     * Warning: might return null!
     *
     * @return RxAdapter Instance
     */
    @Nullable
    public RxAdapter getRxAdapter()
    {
        return mRxAdapter;
    }

    /***
     * Call this if you need access to the Adapter!
     * Warning: might return null!
     *
     * @return RxAdapter Instance
     */
    @Nullable
    public RxAdapterForTypes<DataType> getRxAdapterForTypes()
    {
        return mRxAdapterForTypes;
    }

    /***
     * Call this when you want to bind with multiple Item-Types
     *
     * @param recyclerView RecyclerView instance
     * @param viewHolderInfoList List of ViewHolderInfos
     * @param viewTypeCallback Callback that distinguishes different view Item-Types
     * @return Observable for binding viewHolder
     */
    public Observable<TypesViewHolder<DataType>> bindRecyclerView(
            @NonNull final RecyclerView recyclerView,
            @NonNull final List<ViewHolderInfo> viewHolderInfoList,
            @NonNull final OnGetItemViewType viewTypeCallback)
    {
        mRxAdapter = null;
        mRxAdapterForTypes = new RxAdapterForTypes<>(mDataSet, viewHolderInfoList,
                                                     viewTypeCallback);
        recyclerView.setAdapter(mRxAdapterForTypes);
        return mRxAdapterForTypes.asObservable();
    }

    /***
     * For setting base dataSet
     */
    public RxDataSource<DataType> updateDataSet(List<DataType> dataSet)
    {
        mDataSet = dataSet;
        return this;
    }

    /***
     * For updating Adapter
     */
    public void updateAdapter()
    {
        if (mRxAdapter != null)
        {
            //update the update
            mRxAdapter.updateDataSet(mDataSet);
        }
        else if (mRxAdapterForTypes != null)
        {
            mRxAdapterForTypes.updateDataSet(mDataSet);
        }
    }

    // Transformation methods

    public RxDataSource<DataType> map(Function<? super DataType, ? extends DataType> func)
    {
        mDataSet = Observable.fromIterable(mDataSet).map(func).toList().blockingGet();
        return this;
    }

    public RxDataSource<DataType> filter(Predicate<? super DataType> predicate)
    {
        mDataSet = Observable.fromIterable(mDataSet).filter(predicate).toList().blockingGet();
        return this;
    }

    public final RxDataSource<DataType> last() {
        mDataSet = Collections.singletonList(Observable.fromIterable(mDataSet).blockingLast());
        return this;
    }

    public final RxDataSource<DataType> lastOrDefault(DataType defaultValue) {
        mDataSet = Observable.fromIterable(mDataSet)
                             .takeLast(1)
                             .defaultIfEmpty(defaultValue)
                             .toList()
                             .blockingGet();
        return this;
    }

    public final RxDataSource<DataType> limit(int count) {
        mDataSet = Observable.fromIterable(mDataSet).take(count).toList().blockingGet();
        return this;
    }

    public final RxDataSource<DataType> repeat(final long count)
    {
        List<DataType> dataSet = mDataSet;
        mDataSet = Observable.fromIterable(dataSet).repeat(count).toList().blockingGet();
        return this;
    }

    public RxDataSource<DataType> empty()
    {
        mDataSet = Collections.emptyList();
        return this;
    }

}

File: rxrecycler-adapter/src/main/java/com/minimize/android/rxrecycleradapter/RxDataSource.kt
package com.minimize.android.rxrecycleradapter

import androidx.annotation.LayoutRes
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.Observable
import java.util.Collections.emptyList


/**
 * Created by ahmedrizwan on 26/12/2015.
 */
class RxDataSource<LayoutBinding : ViewDataBinding, DataType>(@LayoutRes private val itemLayout: Int, var dataSet: List<DataType>) {
    /***
     * Call this if you need access to the Adapter!
     * Warning: might return null!
     *
     * @return RxAdapter Instance
     */
    private val rxAdapter: RxAdapter<DataType, LayoutBinding> = RxAdapter(itemLayout, dataSet)

    /***
     * Call this when binding with a single Item-Type
     *
     * @param recyclerView RecyclerView instance
     * @param item_layout Layout id
     * @param <LayoutBinding> ViewDataBinding Type for the layout
     * @return Observable for binding viewHolder
    </LayoutBinding> */
    fun bindRecyclerView(
            recyclerView: RecyclerView
    ): Observable<SimpleViewHolder<DataType, LayoutBinding>> {
        recyclerView.adapter = rxAdapter
        return rxAdapter.asObservable()
    }

    fun asObservable(): Observable<SimpleViewHolder<DataType, LayoutBinding>> {
        return rxAdapter.asObservable()
    }
    /***
     * For setting base dataSet
     */
    fun updateDataSet(dataSet: List<DataType>): RxDataSource<LayoutBinding, DataType> {
        this.dataSet = dataSet
        return this
    }


    /***
     * For setting base dataSet and update adapter
     */
    fun updateDataSet(updatedList: List<DataType>, effectedItem: Int, transactionType: TransactionTypes): RxDataSource<LayoutBinding, DataType> {
        this.dataSet = updatedList
        when (transactionType) {
            TransactionTypes.REPLACE_ALL -> notifyDataSetChanged()
            TransactionTypes.MODIFY -> notifyItemChanged(effectedItem)
            TransactionTypes.DELETE -> notifyItemRemoved(effectedItem)
            TransactionTypes.ADD -> notifyItemInserted(effectedItem)
        }
        return this
    }

    /***
     * For updating Adapter
     */
    fun updateAdapter() {
        //update the update
        notifyDataSetChanged()
    }

    private fun notifyDataSetChanged() {
        //update the update
        rxAdapter.notifyDataSetChanged(dataSet)
    }

    private fun notifyItemChanged(position: Int) {
        //update the update
        rxAdapter.notifyItemChanged(dataSet, position)
    }

    private fun notifyItemRemoved(position: Int) {
        //update the update
        rxAdapter.notifyItemRemoved(dataSet, position)
    }

    private fun notifyItemInserted(position: Int) {
        //update the update
        rxAdapter.notifyItemInserted(dataSet, position)
    }

    // Transformation methods

    fun map(mapper: (DataType) -> DataType): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).map(mapper).toList().blockingGet()
        rxAdapter.updateDataSet(dataSet)
        return this
    }

    fun filter(predicate: (DataType) -> Boolean): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).filter(predicate).toList().blockingGet()
        rxAdapter.updateDataSet(dataSet)
        return this
    }

    fun last(): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).blockingLast())
        return this
    }

    fun first(): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).blockingFirst())
        return this
    }

    fun lastOrDefault(defaultValue: DataType): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet)
                .takeLast(1)
                .defaultIfEmpty(defaultValue)
                .toList()
                .blockingGet()
        return this
    }

    fun limit(count: Int): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).take(count.toLong()).toList().blockingGet()
        return this
    }

    fun repeat(count: Long): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).repeat(count).toList().blockingGet()
        return this
    }

    fun empty(): RxDataSource<LayoutBinding, DataType> {
        dataSet = emptyList<DataType>()
        return this
    }

    fun concatMap(func: (DataType) -> Observable<out DataType>): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).concatMap(func).toList().blockingGet()
        return this
    }

    fun concatWith(observable: Observable<out DataType>): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).concatWith(observable).toList().blockingGet()
        return this
    }

    fun distinct(): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).distinct().toList().blockingGet()
        return this
    }

    fun elementAt(index: Long): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).elementAt(index).blockingGet())
        return this
    }

    fun elementAtOrDefault(index: Long, defaultValue: DataType): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).elementAt(index, defaultValue)
                .blockingGet())
        return this
    }

    fun first(defaultItem: DataType): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).first(defaultItem).blockingGet())
        return this
    }

    fun flatMap(func: (DataType) -> Observable<out DataType>): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).flatMap(func).toList().blockingGet()
        return this
    }

    fun reduce(initialValue: DataType, reducer: (DataType, DataType) -> DataType): RxDataSource<LayoutBinding, DataType> {
        dataSet = listOf(Observable.fromIterable(dataSet).reduce(initialValue, reducer).blockingGet())
        return this
    }

    fun take(count: Long): RxDataSource<LayoutBinding, DataType> {
        dataSet = Observable.fromIterable(dataSet).take(count).toList().blockingGet()
        return this
    }

    companion object {
        const val ALL_ITEMS_EFFECTED = -1
    }

    enum class TransactionTypes {
        REPLACE_ALL,
        DELETE,
        MODIFY,
        ADD
    }

}
